agent/grpc-external/services/dns/server.go

```diff
@@ -0,0 +1,136 @@
+package dns
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/miekg/dns"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+
+	"github.com/hashicorp/consul/proto-public/pbdns"
+)
+
+type Local struct {
+	IP   net.IP
+	Port int
+}
+
+type Config struct {
+	Logger       hclog.Logger
+	DNSServeMux  *dns.ServeMux
+	LocalAddress Local
+}
+
+type Server struct {
+	Config
+}
+
+func NewServer(cfg Config) *Server {
+	return &Server{cfg}
+}
+
+func (s *Server) Register(grpcServer *grpc.Server) {
+	pbdns.RegisterDNSServiceServer(grpcServer, s)
+}
+
+// BufferResponseWriter writes a DNS response to a byte buffer.
+type BufferResponseWriter struct {
+	ResponseBuffer []byte
```

Comment:
Does the buf need to be exported? Do we intend for it to be read/wrote externally?